nan.h

```diff
@@ -2284,11 +2284,14 @@ inline void AsyncExecute (uv_work_t* req) {
  * 2 arguments since node-v0.9.4
  * https://github.com/libuv/libuv/commit/92fb84b751e18f032c02609467f44bfe927b80c5
  */
-inline void AsyncExecuteComplete(uv_work_t* req, int status = 0) {
```

Comment:
doesn't build with g++-8
I wonder if you can take the address of an overloaded function by resolving the fingerprint? I think this works only for direct calls too?

Reply:
https://en.cppreference.com/w/cpp/language/overloaded_address
yes, it should work

Reply:
This is in a header file, so not keeping it inline has the potential to lead to linking errors if an addon happens to include `nan` twice (I know this wouldn’t be a common scenario, but it *is* a realistic one)

I don’t see why keeping it inline wouldn’t work, though, that surprises me – this should generally be possible with GCC 8:

https://godbolt.org/

Reply:
It is not about the inline:
```
/home/mmom/src/node-gdal-async/build/../node_modules/nan/nan.h:2292: multiple definition of `Nan::AsyncExecuteComplete(uv_work_s*, int)'; Debug/obj.target/gdal/src/utils/typed_array.o:/home/mmom/src/node-gdal-async/build/../node_modules/nan/nan.h:2292: first defined here
```

I get a weird error for multiple definitions, both of them on the same line :smile:

Reply:
does anyone understand why the functions need to be inline?

Reply:
The multifile test successfully detects the aforementioned multiple definition problem at the link stage. As @addaleax mentioned, this is a header library. In a header, ordinary, free functions may only be *defined* if marked `inline`, which, in C++ (at least from 98 onwards), has nothing to do with actual inlining of code (that is an implementation optional detail of the compiler, which a modern compiler mostly ignores. In `g++`, I think it might affect the inlinability-score slightly) . What the keyword actually means is that multiple definitions are allowed, with more leeway since C++17.

Reply:
It’s the “realistic scenario” I talked above – if an addon includes `nan.h` twice, i.e. from two different files, then both of those addons will contain definitions for these functions. For `inline` functions that’s allowed, for regular functions not (because one can’t know in advance whether `inline` functions will have an explicit definition after compiling or not, because they might have been, well, inlined :slightly_smiling_face: ).

Reply:
As an addendum, the C++17 thing I mentioned applies to `inline` *variables*, I misremembered that it had anything to do with functions. Note that for non-free functions (member functions), `inline` is implicit if entirely defined in the class declaration. If the definition is outside the class declaration, `inline` is not implicit.